codalab/worker/runtime/kubernetes_runtime.py

```diff
@@ -0,0 +1,234 @@
+import json
+import logging
+from dateutil import tz
+from typing import Any, Dict, Optional, Tuple
+from urllib3.exceptions import MaxRetryError, NewConnectionError  # type: ignore
+
+from kubernetes import client, utils  # type: ignore
+from kubernetes.utils.create_from_yaml import FailToCreateError  # type: ignore
+from kubernetes.client.rest import ApiException
+
+from codalab.worker.docker_utils import DEFAULT_RUNTIME
+from codalab.common import BundleRuntime
+from codalab.worker.runtime import Runtime
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+# https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#create-pod-v1-core
+
+removeprefix = lambda l, p: l[len(p) :]
+
+
+class KubernetesRuntime(Runtime):
+    """Runtime that launches Kubernetes pods."""
+
+    @property
+    def name(self) -> str:
+        return BundleRuntime.KUBERNETES.value
+
+    def __init__(self, work_dir: str, auth_token: str, cluster_host: str, cert_path: str):
+        # Configure and initialize Kubernetes client
+        self.work_dir = work_dir
+
+        # TODO: Unify this code with the client setup steps in kubernetes_worker_manager.py
+        configuration: client.Configuration = client.Configuration()
+        configuration.api_key_prefix['authorization'] = 'Bearer'
+        configuration.api_key['authorization'] = auth_token
+        configuration.host = cluster_host
+        configuration.ssl_ca_cert = cert_path
+        if configuration.host == "https://codalab-control-plane:6443":
+            # Don't verify SSL if we are connecting to a local cluster for testing / development.
+            configuration.verify_ssl = False
+            configuration.ssl_ca_cert = None
+            del configuration.api_key_prefix['authorization']
+            del configuration.api_key['authorization']
+            configuration.debug = False
+
+        self.k8_client: client.ApiClient = client.ApiClient(configuration)
+        self.k8_api: client.CoreV1Api = client.CoreV1Api(self.k8_client)
+
+    def get_nvidia_devices(self, use_docker=True):
+        """
+        Returns a Dict[index, UUID] of all NVIDIA devices available to docker
+
+        Arguments:
+            use_docker: whether or not to use a docker container to run nvidia-smi.
+
+        Raises docker.errors.ContainerError if GPUs are unreachable,
+            docker.errors.ImageNotFound if the CUDA image cannot be pulled
+            docker.errors.APIError if another server error occurs
+        """
+        return {}
+
+    def get_container_ip(self, network_name: str, pod_name: str):
+        """Returns (finished boolean, exitcode or None of bundle, failure message or None)"""
+        pod = self.k8_api.read_namespaced_pod_status(pod_name, "default")
+        return pod.status.pod_ip
+
+    def start_bundle_container(
+        self,
+        bundle_path,
+        uuid,
+        dependencies,
+        command,
+        docker_image,
+        network=None,
+        cpuset=None,
```

Comment:
We don't need to heed cpuset and gpuset anymore, right? Because work may be on a separate machine?